pennylane/utils.py

```diff
@@ -98,5 +109,140 @@ def unflatten(flat, model):
     # pylint:disable=len-as-condition
     res, tail = _unflatten(np.asarray(flat), model)
     if len(tail) != 0:
-        raise ValueError('Flattened iterable has more elements than the model.')
+        raise ValueError("Flattened iterable has more elements than the model.")
     return res
+
+
+def _inv_dict(d):
+    """Reverse a dictionary mapping.
+
+    Returns multimap where the keys are the former values,
+    and values are sets of the former keys.
+
+    Args:
+        d (dict[a->b]): mapping to reverse
+
+    Returns:
+        dict[b->set[a]]: reversed mapping
+    """
+    ret = {}
+    for k, v in d.items():
+        ret.setdefault(v, set()).add(k)
+    return ret
+
+
+def _get_default_args(func):
+    """Get the default arguments of a function.
+
+    Args:
+        func (function): a valid Python function
+
+    Returns:
+        dict: dictionary containing the argument name and tuple
+        (positional idx, default value)
+    """
+    signature = inspect.signature(func)
+    return {
+        k: (idx, v.default)
+        for idx, (k, v) in enumerate(signature.parameters.items())
+        if v.default is not inspect.Parameter.empty
+    }
+
+
+def expand(U, wires, num_wires):
+    r"""Expand a multi-qubit operator into a full system operator.
+
+    Args:
+        U (array): :math:`2^n \times 2^n` matrix where n = len(wires).
+        wires (Sequence[int]): Target subsystems (order matters! the
+            left-most Hilbert space is at index 0).
+
+    Returns:
+        array: :math:`2^N\times 2^N` matrix. The full system operator.
+    """
+    if num_wires == 1:
+        # total number of wires is 1, simply return the matrix
+        return U
+
+    N = num_wires
+    wires = np.asarray(wires)
+
+    if np.any(wires < 0) or np.any(wires >= N) or len(set(wires)) != len(wires):
+        raise ValueError("Invalid target subsystems provided in 'wires' argument.")
+
+    if U.shape != (2 ** len(wires), 2 ** len(wires)):
+        raise ValueError("Matrix parameter must be of size (2**len(wires), 2**len(wires))")
+
+    # generate N qubit basis states via the cartesian product
+    tuples = np.array(list(itertools.product([0, 1], repeat=N)))
+
+    # wires not acted on by the operator
+    inactive_wires = list(set(range(N)) - set(wires))
+
+    # expand U to act on the entire system
+    U = np.kron(U, np.identity(2 ** len(inactive_wires)))
+
+    # move active wires to beginning of the list of wires
+    rearranged_wires = np.array(list(wires) + inactive_wires)
```

Comment:
Yep. This is definitely a bottleneck in the current implementation - in order to compute the shared eigenbasis, each generator is being expanded to a 2^N x 2^N matrix to be used with `sp.linalg.eigh`.

There is probably a better way of doing this - perhaps a way of doing this analytically using the properties of Pauli matrices?

Note that this process is only performed once, on the first optimization step, and the resulting eigenbasis is saved for future use.

Reply:
If the generator has eigenvalues (a,b,c,...), won't expanding it just preserve the eigenvalues, but increase their multiplicity?
e.g., the eigenvalues of X are (-1,+1), and the eigenvalues of X\otimes\Identity are (-1,-1, +1, +1)

Eigenvectors should also have a clear pattern

Reply:
>  e.g., the eigenvalues of X are (-1,+1), and the eigenvalues of X\otimes\Identity are (-1,-1, +1, +1)

That's a very good point, I'll need to think more about this.

`np.kron` appears to work well using vectors instead of matrices:
```python
>>> np.kron(np.array([-1,1]), np.array([-1, 1]))
array([ 1, -1, -1,  1])
```